batch/batch/semaphore.py

```diff
@@ -23,24 +24,33 @@ def __init__(self, sem, weight):
         await self.sem.acquire(self.weight)
 
     async def __aexit__(self, exc_type, exc, tb):
-        await self.sem.release(self.weight)
+        self.sem.release(self.weight)
 
 
-class WeightedSemaphore:
+class FIFOWeightedSemaphore:
     def __init__(self, value=1):
         self.value = value
-        self.cond = asyncio.Condition()
+        self.queue = collections.deque()
 
     async def acquire(self, weight):
-        while self.value < weight:
-            async with self.cond:
-                await self.cond.wait()
-        self.value -= weight
+        if not self.queue and self.value >= weight:
```

Comment:
This queue jumping is still an issue.

Reply:
Can you be more specific? I thought I fixed this. If the queue is empty (`not self.queue`) and we have the capacity to run, then we should run. Otherwise, we should add an event to the queue and wait for resources to free.

I agree if we didn't check if there are items in the queue, then we would be jumping jobs. However, if you don't have this if-statement block of code, then you'll deadlock trying to acquire the semaphore for the first item.

Reply:
Ah, if if the queue isn't empty.  I guess I'm wondering, why even have this test?  Why not just always put it on the queue?

Reply:
> you'll deadlock trying to acquire the semaphore for the first item

Ah, I'd missed that, thanks.